fabric-ingredient-api-v1/src/main/java/net/fabricmc/fabric/mixin/ingredient/IngredientMixin.java

```diff
@@ -0,0 +1,125 @@
+
+import net.fabricmc.fabric.api.ingredient.v1.CustomIngredient;
+import net.fabricmc.fabric.api.ingredient.v1.CustomIngredientSerializer;
+import net.fabricmc.fabric.api.ingredient.v1.FabricIngredient;
+import net.fabricmc.fabric.impl.ingredient.CustomIngredientImpl;
+import net.fabricmc.fabric.impl.ingredient.builtin.OrIngredient;
+
+@Mixin(value = Ingredient.class, priority = 900) // overwrite before faux ingredient extension API injects in the method
+public class IngredientMixin implements FabricIngredient {
+	/**
+	 * Inject right when vanilla detected a json object and check for our custom key.
+	 */
+	@Inject(
+			at = @At(
+					value = "INVOKE",
+					target = "net/minecraft/recipe/Ingredient.entryFromJson (Lcom/google/gson/JsonObject;)Lnet/minecraft/recipe/Ingredient$Entry;",
+					ordinal = 0
+			),
+			method = "fromJson",
+			cancellable = true
+	)
+	private static void injectFromJson(JsonElement json, CallbackInfoReturnable<Ingredient> cir) {
+		JsonObject obj = json.getAsJsonObject();
+
+		if (obj.has(CustomIngredientImpl.TYPE_KEY)) {
+			Identifier id = new Identifier(JsonHelper.getString(obj, CustomIngredientImpl.TYPE_KEY));
+			CustomIngredientSerializer<?> serializer = CustomIngredientSerializer.get(id);
+
+			if (serializer != null) {
+				cir.setReturnValue(serializer.read(obj).toVanilla());
+			} else {
+				throw new IllegalArgumentException("Unknown custom ingredient type: " + id);
+			}
+		}
+	}
+
+	/**
+	 * Throw exception when someone attempts to use our custom key inside an array ingredient.
+	 * The {@link OrIngredient} should be used instead.
+	 */
+	@Inject(at = @At("HEAD"), method = "entryFromJson")
+	private static void injectEntryFromJson(JsonObject obj, CallbackInfoReturnable<?> cir) {
+		if (obj.has(CustomIngredientImpl.TYPE_KEY)) {
+			throw new IllegalArgumentException("Custom ingredient cannot be used inside an array ingredient. You can replace the array by a fabric:or ingredient.");
+		}
+	}
+
+	@Inject(
+			at = @At(
+					value = "INVOKE",
+					target = "net/minecraft/network/PacketByteBuf.writeCollection (Ljava/util/Collection;Lnet/minecraft/network/PacketByteBuf$PacketWriter;)V"
+			),
+			method = "write",
+			cancellable = true
+	)
+	@SuppressWarnings({"unchecked", "rawtypes"})
+	private void injectWrite(PacketByteBuf buf, CallbackInfo ci) {
+		// Done here instead of writing the ingredient in the subclass for compat with faux ingredient extension API
+		CustomIngredient customIngredient = getCustomIngredient();
+
+		if (customIngredient != null) {
+			buf.writeVarInt(CustomIngredientImpl.PACKET_MARKER);
+			buf.writeIdentifier(customIngredient.getSerializer().getIdentifier());
+			((CustomIngredientSerializer) customIngredient.getSerializer()).write(buf, customIngredient);
+			ci.cancel();
+		}
+	}
+
+	/**
+	 * @author FabricMC
+	 * @reason Support custom ingredient network deserialization.
+	 */
+	@Overwrite
+	public static Ingredient fromPacket(PacketByteBuf buf) {
```

Comment:
This could be replaced with an injection

Reply:
Could be but I don't see the point since I'm basically overwriting the entire method.

Reply:
You could allow vanilla code to run, for better compat for things that inject there. And for vanilla code reading that, you can set packbufers index to one before reading

Reply:
So I can "unread" the int? That would work better indeed.

Reply:
Yep, just store old index before reading and restore it if size is different than -1

Reply:
Done: https://github.com/FabricMC/fabric/pull/2586/commits/9a154dafb484bca11ed4f243b36758756c356ac2.